articles/112_generated_interfaces_cpp.md

```diff
@@ -0,0 +1,119 @@
+* namespace for ROS messages: `package_name::msg`.
+* namespace for ROS services: `package_name::srv`.
+
+
+## Generated files
+
+Following the C++ style guide of ROS 2 the namespace hierarchy is mapped to a folder structure.
+The filenames use lowercase alphanumeric characters with underscores for separating words and end with either `.hpp` or `.cpp`.
+
+
+## Messages
+
+For a message a templated `struct` with the same name followed by an underscore is generated.
+The single template argument is the allocator for the data structure.
+
+For ease of use there is a `typedef` with the same name as the message which uses a default allocator (e.g. `std::allocator`).
+
+For each message two files are being generated:
+
+* `<my_message_name>.hpp` currently only includes `<my_message_name>__struct.hpp`
+* `<my_message_name>__struct.hpp` containing the definition of the struct
+
+This allow to add addition files beside the one with the suffix `__struct` to provide additional functionality.
+For each additional functionality it can be decided to include it from the first header file.
+
+
+### Members
+
+The struct has same-named public member variables for every field of the message.
+For each field a `typedef` is created which is named after the member with a leading underscore and a trailing `_type`.
+
+
+### Constants
+
+Numeric constants are defined as `enums` within the struct.
+
+All other constants are declared as `static const` members in the struct and their values are defined outside of the struct.
+
+
+### Constructors
+
+The *default constructor* initializes all members with their default value.
+Optionally the constructor can be invoked with an allocator.
+
+The struct has no constructor with positional arguments for the members.
+The short reason for this is that if code would rely on positional arguments to construct data objects changing a message definition would break existing code in various ways.
+Since this would discourage evolution of message definitions the data structures should be populated by setting the members separately, e.g. using the setter methods.
+
+
+### Setters
+
+For each field a *setter* method is generated to enable (method chaining)[https://isocpp.org/wiki/faq/ctors#named-parameter-idiom].
+They are named after the fields with a leading `set__`.
+The setter methods have a single argument to pass the value for the member variable.
+Each setter method returns the struct itself.
+
+
+### Operators
+
+The comparison operators `==` and `!=` perform the comparison on a per member basis.
+
+
+### Pointer types
+
+The struct contains `typedefs` for the four common pointer types `plain pointer`, `std::shared_ptr`, `std::unique_ptr`, `std::weak_ptr`.
+For each pointer type there a non-const and a const `typedef`:
+
+* `Ptr` and `ConstPtr`
+* `SharedPtr` and `ConstSharedPtr`
+* `UniquePtr` and `ConstUniquePtr`
+* `WeakPtr` and `ConstWeakPtr`
```

Comment:
I'd say no; `auto_ptr` is deprecated in C++11. The others look fine. People might find it weird that `Ptr` is now a "raw" pointer rather than a shared pointer, as it is in ROS 1. Not sure what else we could do though. Maybe `RawPtr` and `ConstRawPtr`, but I don't really like that either.